python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -896,6 +896,116 @@ def corr(self, method="pearson", min_periods=1):
+        >>> import cudf
+        >>> gdf = cudf.DataFrame({
+        ...             "id": ["a", "a", "a", "b", "b", "b", "c", "c", "c"],
+        ...             "val1": [5, 4, 6, 4, 8, 7, 4, 5, 2],
+        ...             "val2": [4, 5, 6, 1, 2, 9, 8, 5, 1],
+        ...             "val3": [4, 5, 6, 1, 2, 9, 8, 5, 1]})
+        >>> gdf
+        id  val1  val2  val3
+        0  a     5     4     4
+        1  a     4     5     5
+        2  a     6     6     6
+        3  b     4     1     1
+        4  b     8     2     2
+        5  b     7     9     9
+        6  c     4     8     8
+        7  c     5     5     5
+        8  c     2     1     1
+        >>> gdf.groupby("id").cov()
+                    val1       val2       val3
+        id
+        a  val1  1.000000   0.500000   0.500000
+           val2  0.500000   1.000000   1.000000
+           val3  0.500000   1.000000   1.000000
+        b  val1  4.333333   3.500000   3.500000
+           val2  3.500000  19.000000  19.000000
+           val3  3.500000  19.000000  19.000000
+        c  val1  2.333333   3.833333   3.833333
+           val2  3.833333  12.333333  12.333333
+           val3  3.833333  12.333333  12.333333
+        """
+
+        # create expanded dataframe consisting all combinations of the
+        # struct columns-pairs to be correlated
+        # i.e (('col1', 'col1'), ('col1', 'col2'), ('col2', 'col2'))
+        column_names = self.grouping.values.columns.tolist()
+        len_cols = len(column_names)
+
+        column_pair_structs = {}
+        for x, y in itertools.combinations_with_replacement(column_names, 2):
+            column_pair_structs[(x, y)] = cudf.DataFrame._from_data(
+                {"x": self.obj._data[x], "y": self.obj._data[y]}
+            ).to_struct()
+        column_pair_groupby = cudf.DataFrame._from_data(
+            column_pair_structs
+        ).groupby(by=self.grouping.keys)
+
+        try:
+            gb_cov = column_pair_groupby.agg(
+                lambda x: x.cov(min_periods, ddof)
+            )
+        except RuntimeError as e:
+            if "Unsupported groupby reduction type-agg combination" in str(e):
+                raise TypeError(
+                    "Covariance accepts only numerical column-pairs"
+                )
+            raise
+        # ensure that column-pair labels are arranged in ascending order
+        cols_list = [
+            (y, x) if i > j else (x, y)
+            for j, y in enumerate(column_names)
+            for i, x in enumerate(column_names)
+        ]
+        cols_split = [
+            cols_list[i : i + len_cols]
+            for i in range(0, len(cols_list), len_cols)
+        ]
+
+        # interleave: combine the correlation results for each column-pair
+        # into a single column
+        res = cudf.DataFrame._from_data(
+            {
+                x: gb_cov.loc[:, i].interleave_columns()
+                for i, x in zip(cols_split, column_names)
+            }
+        )
```

Comment:
It's debatable where we should put `combine_columns`, but the idea is that we don't resort to `.loc` since it's expensive.

```suggestion
        def combine_columns(gb_cov, ys):
            list_of_columns = list(map(gb_cov._data.__getitem__, ys))
            frame = cudf.core.frame.Frame._from_columns(list_of_columns, ys)
            return interleave_columns(frame)

        # interleave: combine the correlation results for each column-pair
        # into a single column
        res = cudf.DataFrame._from_data(
            {
                x: combine_columns(gb_cov, ys)
                for ys, x in zip(cols_split, column_names)
            }
        )

```

Reply:
Minor note: I think using `x`, `y` to denote the two operands participating in the covariance compute should be standardized through out the PR, otherwise it's pretty hard to figure out what the code is referring to when reading it. (`i` in the previous code e.g.).

Reply:
Simplifying the suggestion:
```diff
- list_of_columns = list(map(gb_cov._data.__getitem__, ys))
+ list_of_columns = [gb_cov._data[y] for y in ys]
```